src/docker/local.rs

```diff
@@ -1,56 +1,46 @@
 use std::io;
-use std::path::Path;
 use std::process::ExitStatus;
 
-use super::engine::*;
 use super::shared::*;
-use crate::cargo::CargoMetadata;
 use crate::errors::Result;
 use crate::extensions::CommandExt;
 use crate::file::{PathExt, ToUtf8};
 use crate::shell::{MessageInfo, Stream};
-use crate::{Config, Target};
 use eyre::Context;
 
-#[allow(clippy::too_many_arguments)] // TODO: refactor
 pub(crate) fn run(
-    engine: &Engine,
-    target: &Target,
+    options: DockerOptions,
+    paths: DockerPaths,
     args: &[String],
-    metadata: &CargoMetadata,
-    config: &Config,
-    uses_xargo: bool,
-    sysroot: &Path,
-    msg_info: MessageInfo,
-    docker_in_docker: bool,
-    cwd: &Path,
+    msg_info: &mut MessageInfo,
 ) -> Result<ExitStatus> {
-    let mount_finder = MountFinder::create(engine, docker_in_docker)?;
-    let dirs = Directories::create(&mount_finder, metadata, cwd, sysroot)?;
+    let dirs = paths.directories();
 
-    let mut cmd = cargo_safe_command(uses_xargo);
+    let mut cmd = cargo_safe_command(options.uses_xargo());
     cmd.args(args);
 
-    let mut docker = subcommand(engine, "run");
+    let mut docker = subcommand(options.engine(), "run");
     docker_userns(&mut docker);
-    docker_envvars(&mut docker, config, target, msg_info)?;
+    docker_envvars(&mut docker, options.config(), options.target(), msg_info)?;
 
     let mount_volumes = docker_mount(
         &mut docker,
-        metadata,
-        &mount_finder,
-        config,
-        target,
-        cwd,
+        &options,
+        &paths,
         |docker, val| mount(docker, val, ""),
         |_| {},
     )?;
 
     docker.arg("--rm");
 
-    docker_seccomp(&mut docker, engine.kind, target, metadata)
-        .wrap_err("when copying seccomp profile")?;
-    docker_user_id(&mut docker, engine.kind);
+    docker_seccomp(
+        &mut docker,
```

Comment:
I've tried not to overly rely on just having the structs passed: if there's any question of if the logic could need to be accessed outside and the logic does not rely on most of the members or it would require a lot of work to create those members, I've avoided passing the struct itself.